src/librustc_mir_build/build/block.rs

```diff
@@ -215,6 +215,8 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 assert_eq!(self.push_unsafe_count, 0);
                 match self.unpushed_unsafe {
                     Safety::Safe => {}
+                    // no longer treat `unsafe fn`s as `unsafe` contexts (see RFC #2585)
+                    Safety::FnUnsafe if self.hir.tcx().features().unsafe_block_in_unsafe_fn => {}
```

Comment:
This is the wrong trigger. Even when the feature is enabled, unless `unsafe_op_in_unsafe_fn` is at least "warn", these unsafe blocks are *not* unused.

Reply:
In other words, you did not properly implement this part from the RFC:

> Moreover, if and only if the unsafe_op_in_unsafe_fn lint is not allowed, we no longer warn that an unsafe block is unnecessary when it is nested immediately inside an unsafe fn. 

That also explains why you didn't test for the RFC'd behavior in these cases (see my comments on the tests).

Reply:
Oh! Alright, I didn't remember that part. Sorry about that 😅